beacon-chain/sync/rpc_data_column_sidecars_by_root.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/OffchainLabs/prysm/v6/consensus-types/primitives"
+	"github.com/OffchainLabs/prysm/v6/monitoring/tracing"
+	"github.com/OffchainLabs/prysm/v6/monitoring/tracing/trace"
+	"github.com/OffchainLabs/prysm/v6/time/slots"
+	libp2pcore "github.com/libp2p/go-libp2p/core"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var notDataColumnsByRootIdentifiersError = errors.New("not data columns by root identifiers")
+
+// dataColumnSidecarByRootRPCHandler handles the data column sidecars by root RPC request.
+// https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/p2p-interface.md#datacolumnsidecarsbyroot-v1
+func (s *Service) dataColumnSidecarByRootRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	ctx, span := trace.StartSpan(ctx, "sync.dataColumnSidecarByRootRPCHandler")
+	defer span.End()
+
+	batchSize := flags.Get().DataColumnBatchLimit
+	numberOfColumns := params.BeaconConfig().NumberOfColumns
+
+	// Check if the message type is the one expected.
+	ref, ok := msg.(*types.DataColumnsByRootIdentifiers)
+	if !ok {
+		return notDataColumnsByRootIdentifiersError
+	}
+
+	requestedColumnIdents := *ref
+	remotePeerId := stream.Conn().RemotePeer()
+
+	ctx, cancel := context.WithTimeout(ctx, ttfbTimeout)
+	defer cancel()
+
+	SetRPCStreamDeadlines(stream)
+
+	// Penalize peers that send invalid requests.
+	if err := validateDataColumnsByRootRequest(requestedColumnIdents); err != nil {
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(remotePeerId)
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		return errors.Wrap(err, "validate data columns by root request")
+	}
+
+	requestedColumnsByRoot := make(map[[fieldparams.RootLength]byte][]uint64)
+	for _, columnIdent := range requestedColumnIdents {
+		var root [fieldparams.RootLength]byte
+		copy(root[:], columnIdent.BlockRoot)
+		requestedColumnsByRoot[root] = append(requestedColumnsByRoot[root], columnIdent.Columns...)
+	}
+
+	// Sort by column index for each root.
+	for _, columns := range requestedColumnsByRoot {
+		slices.Sort(columns)
+	}
+
+	// Format nice logs.
+	requestedColumnsByRootLog := make(map[string]interface{})
+	for root, columns := range requestedColumnsByRoot {
+		rootStr := fmt.Sprintf("%#x", root)
+		requestedColumnsByRootLog[rootStr] = "all"
+		if uint64(len(columns)) != numberOfColumns {
+			requestedColumnsByRootLog[rootStr] = columns
+		}
+	}
+
+	// Compute the oldest slot we'll allow a peer to request, based on the current slot.
+	minReqSlot, err := dataColumnsRPCMinValidSlot(s.cfg.clock.CurrentSlot())
+	if err != nil {
+		return errors.Wrapf(err, "data columns RPC min valid slot")
+	}
+
+	log := log.WithFields(logrus.Fields{
+		"peer":    remotePeerId,
+		"columns": requestedColumnsByRootLog,
+	})
+
+	defer closeStream(stream, log)
+
+	var ticker *time.Ticker
+	if len(requestedColumnIdents) > batchSize {
+		ticker = time.NewTicker(time.Second)
```

Comment:
this should have some kind of `defer ticker.Stop() ` right

Reply:
Was required before Go 1.23.

Go doc of `NewTicker`:

> Before Go 1.23, the garbage collector did not recover tickers that had not yet expired or been stopped, so code often immediately deferred t.Stop after calling NewTicker, to make the ticker recoverable when it was no longer needed. As of Go 1.23, the garbage collector can recover unreferenced tickers, even if they haven't been stopped. The Stop method is no longer necessary to help the garbage collector. (Code may of course still want to call Stop to stop the ticker for other reasons.)